### PR TITLE
avoidance failsafe in commander

### PR DESCRIPTION
**Pull request by mrivi**
**Describe problem solved by the proposed pull request**
I would like to unify the failsafe action with the communication to the user for the obstacle avoidance. Right now the failsafe happens in the the mc_pos_control issuing a vehicle command. 

EDIT: another reason why I would like to make this change is that there is a refactor of the avoidance interface in the works ( #11445). The setpoints coming from the avoidance will be injected directly into the flight task (AutoMapper and Manual Position) therefore the subscription to `vehicle_trajectory_waypoint` and check for validity doesn't need to be anymore in the mc_pos_control. 

EDIT2: the PR now refactors the failsafe to be handled differently according to the navigation states in which the avoidance is enabled. Here's a summary of the desired behavior 

| Nav State  | Failsafe |
| ------------- | ------------- |
| Mission  | Loiter  |
| Takeoff  | Land  |
| Land  | OA not available  |
| Loiter  | OA not available  |
| RTL  | disabled OA and stay in RTL |
| Follow Target  | OA not available  |

**Test data / coverage**
SITL tested

### CODE REVIEW

File: src/modules/commander/state_machine_helper.cpp [hunk 1]
```diff
@@ -568,6 +568,8 @@ bool set_nav_state(vehicle_status_s *status, actuator_armed_s *armed, commander_
 
 		} else if (is_armed && check_invalid_pos_nav_state(status, old_failsafe, mavlink_log_pub, status_flags, false, true)) {
 			// nothing to do - everything done in check_invalid_pos_nav_state
+		} else if (status->avoidance_lost) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER;
```
Comment (mrivi) [inline src/modules/commander/state_machine_helper.cpp]: Not sure about this failsafe because RTL could be triggered by low battery etc. Should the vehicle loiter or land?

File: src/modules/commander/Commander.hpp [hunk 2]
```diff
@@ -155,27 +168,27 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
...
+	int		_telemetry_status_sub[ORB_MULTI_MAX_INSTANCES] {};
+
+	uint64_t	_datalink_last_heartbeat_gcs{0};
+	uint64_t	_datalink_lost{0};
+
+	uint64_t	_datalink_last_heartbeat_onboard_controller{0};
+	uint64_t	_onboard_controller_lost{0};
+
+	uint64_t	_datalink_last_heartbeat_avoidance_system{0};
+	uint64_t	_avoidance_system_lost{0};
```
Comment (cmic0) [inline src/modules/commander/Commander.hpp]: I think that you can use a bool type for _avoidance_system_lost

Comment (cmic0) [reply src/modules/commander/Commander.hpp]: I did it [here](https://github.com/PX4/Firmware/pull/11456/commits/6e1e6083adb9190e158dc547f277f1aa49d2798c)

**Comment by mrivi** [comment]
New failsafe logic

| Nav State  | Failsafe |
| ------------- | ------------- |
| Mission  | Loiter  |
| Takeoff  | Land  |
| Land  | OA not available  |
| Loiter  | OA not available  |
| RTL  | disabled OA and stay in RTL |
| Follow Target  | OA not available  |

**Comment by MaEtUgR** [comment]
> RTL | disabled OA and stay in RTL
> -- | --

Don't you use/test OA in RTL? And if the OA is suddenly not available anymore you crash into your the obstacle if you continue? Just trying to help and think of corner cases.

**Comment by mrivi** [comment]
@MaEtUgR If the avoidance is healthy it runs during RTL. However if the connection to the companion is lost or the avoidance node dies, we should continue with RTL expecially if the cause of the RTL is already a failsafe (e.g. low battery). Otherwise if we put the vehicle to hold and the battery drains, the vehicle would fall down. 
We tough that switching off the avoidance with some clear and very visible warning in QGC would be more safe. Toughts?

**Comment by MaEtUgR** [comment]
@mrivi Yes makes sense. Another option would be to just land where it currently is to not horizontally run into any obstacles but this still assumes there's no obstacle or inappropriate terrain underneath...

**Comment by jkflying** [comment]
This is out of date and technically fixed, although it is still a bit awkward that the message to the user isn't sent from the same place as the actual failsafe decision is made.
